pallets/utxo/src/rewards.rs

```diff
@@ -0,0 +1,102 @@
+where
+    T: Config + pallet_authorship::Config,
+{
+    fn note_author(author: T::AccountId) {
+        match convert_to_h256::<T>(&author) {
+            Ok(author_h256) => {
+                // store the block author. Reward during the `fn finalize()` phase.
+                <BlockAuthor<T>>::put(author_h256);
+            }
+            Err(e) => {
+                log::warn!("failed to find author: {:?}", e);
+            }
+        }
+    }
+
+    fn note_uncle(_author: T::AccountId, _age: T::BlockNumber) {
+        log::info!("TODO: no support for this. Or is there...?");
+    }
+}
+
+/// Returns the newly reduced reward amount for a Block Author.
+/// How much a reward is reduced, is based on the config's`RewardReductionFraction`.
+fn get_reward_amount<T: Config>(block_number: T::BlockNumber) -> u128 {
+    let reduction_fraction = T::RewardReductionFraction::get();
+    let reduction_period: T::BlockNumber = T::RewardReductionPeriod::get();
+    let mut reward_amount = T::InitialReward::get();
+
+    if let Some(mut counter) = block_number.checked_div(&reduction_period) {
+        // how many times the initial reward should be slashed.
+        while counter > T::BlockNumber::zero() {
+            counter = counter.checked_sub(&T::BlockNumber::one()).unwrap_or(T::BlockNumber::zero());
+
+            reward_amount = reward_amount
+                .checked_sub(reduction_fraction.mul_ceil(reward_amount))
+                .unwrap_or(1); // TODO: this is only testnet specific to reward at least 1
+
+            if reward_amount.is_zero() {
+                // TODO: this is only testnet specific to reward at least 1
+                return 1;
+            }
+        }
+    }
+
+    reward_amount
+}
+
+/// Rewards the block author with a utxo of value based on the `BlockAuthorRewardAmount`
+/// and the transaction fees.
+pub(crate) fn reward_block_author<T: Config>(block_number: T::BlockNumber) {
+    let fees_total = <RewardTotal<T>>::take();
+    if let Some(reward_amount) = get_reward_amount::<T>(block_number).checked_add(fees_total) {
+        // give rewards only if a block author is found
+        if let Some(block_author) = <BlockAuthor<T>>::take() {
+            let utxo = TransactionOutput::new_pubkey(reward_amount, block_author.clone());
+
+            let hash = {
+                let b_num = block_number.saturated_into::<u64>();
+                BlakeTwo256::hash_of(&(&utxo, b_num, "author_reward"))
+            };
+
+            if !<UtxoStore<T>>::contains_key(hash) {
+                <UtxoStore<T>>::insert(hash, utxo.clone());
+
+                <Pallet<T>>::deposit_event(Event::<T>::BlockAuthorRewarded(utxo));
+            }
+        } else {
+            log::warn!("no block author found for block number {:?}", block_number);
+            <RewardTotal<T>>::put(fees_total + reward_amount);
+        }
+    } else {
+        log::warn!("problem adding the block author reward and the fees.");
```

Comment:
Talking about the `get_reward_amount`, I think it should never fail... otherwise the node should panic. There has to be tests testing the full input space of this function to ensure it'll never fail. Which also means that it has to be simple enough. This is why I'm not happy with that loop in there... I seem to fail to understand the purpose of that loop.

Reply:
Given that this should never fail, I think the right behavior is to just `panic!`. You should use `expect` to achieve that.